host/fabricshared/ProtocolGateway.Host.Fabric.FabricShared.Configuration/ConfigurationProvider.cs

```diff
@@ -0,0 +1,402 @@
+﻿namespace ProtocolGateway.Host.Fabric.FabricShared.Configuration
+{
+    #region Using Clauses
+    using System;
+    using System.Text;
+    using System.Collections.ObjectModel;
+    using System.Fabric;
+    using System.Fabric.Description;
+    using System.IO;
+    using System.Security;
+    using System.Security.Cryptography;
+    using System.ComponentModel;
+    using Logging;
+    using Properties;
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Converters;
+    using Security;
+    #endregion
+
+    /// <summary>
+    /// Generic configuration provider.
+    /// </summary>
+    public sealed class ConfigurationProvider<TConfiguration> : IConfigurationProvider<TConfiguration> where TConfiguration : class
+    {
+        #region Constants
+        /// <summary>
+        /// Name of the configuration package object.
+        /// </summary>
+        const string ConfigurationPackageObjectName = @"Config";
+
+        /// <summary>
+        /// The component name for debug and diagnostics messages
+        /// </summary>
+        const string ComponentName = @"Configuration Provider";
+        #endregion
+        #region Events
+        /// <summary>
+        /// Configuration changed event for property based configuration changes.
+        /// </summary>
+        public event EventHandler OnConfigurationPropertyChangedEvent;
+
+        /// <summary>
+        /// Configuration changed event for property based configuration changes.
+        /// </summary>
+        public event EventHandler<ConfigurationChangedEventArgs<TConfiguration>> OnConfigurationClassChangedEvent;
```

Comment:
Done but typically I believe most people in .NET expect the events to prefix with On.